### PR TITLE
MIDI-43: Variational Autoencoder

### PR DESCRIPTION
**Pull request by SamuelJanas**
[MIDI-43](https://sacrebleu.atlassian.net/browse/MIDI-43)
[MIDI-45](https://sacrebleu.atlassian.net/browse/MIDI-45)
[MIDI-46](https://sacrebleu.atlassian.net/browse/MIDI-46)

[MIDI-43]: https://sacrebleu.atlassian.net/browse/MIDI-43?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MIDI-45]: https://sacrebleu.atlassian.net/browse/MIDI-45?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MIDI-46]: https://sacrebleu.atlassian.net/browse/MIDI-46?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SamuelJanas** [comment]
`python train.py --config-name="vae_config.yaml`
### Interpolations
Interpolations tend to happen 'through' another number, which is a good sign (1-5 interpolation for clear image) meaning that the variational autoencoder is likely working.
interpolations_VAE_MNIST

### Sample generations
Generating capabilities of AE aren't the best. We see a big improvement using VAE instead:
**AE**: 
sample_AE_MNIST

**VAE**:
sample_VAE_MNIST
Overall, the VAE is a big improvement over the AE in both interpolation and sample generation. 🎉

**Comment by SamuelJanas** [comment]
Tested VAE embeddings on random forest classifier:
## Visualization:
reconstructions_VAE_MNIST
That's how the embeddings look like when we transform them into 2D, note the padding at the end to fill the square.
## Accuracy
- Accuracy: 94.76%
It's slightly lower than accuracy on vanilla AE with the same embedding size. The results makes sense I believe, since classification doesn't benefit from the added variation as that change is strictly towards the generative part of the task.

**Comment by SamuelJanas** [comment]
>For the Random Forest part of this experiment, can you prepare a plot showing the validation accuracy on MNIST as a function of the embedding size? You could check for the embedding sizes from range(1, 32) - I think this could be an interesting exercise 

Trained 32 models for the best accuracy, here are the results:
accuracy vs embedding
Around embedding size 5 we get ~94% accuracy. This value does not change no matter the embedding size.

File: models/VQVAE.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        # Exponential moving average of the embeddings.
+        self.m_i_ts = SonnetExponentialMovingAverage(decay, e_i_ts.shape)
+
+    def forward(self, x):
+        flat_x = x.permute(0, 2, 3, 1).reshape(-1, self.embedding_dim)
+        distances = (flat_x**2).sum(1, keepdim=True) - 2 * flat_x @ self.e_i_ts + (self.e_i_ts**2).sum(0, keepdim=True)
+        encoding_indices = distances.argmin(1)
+        quantized_x = F.embedding(encoding_indices.view(x.shape[0], *x.shape[2:]), self.e_i_ts.transpose(0, 1)).permute(
+            0, 3, 1, 2
+        )
```
Comment (JasiekKaczmarczyk) [inline models/VQVAE.py:238]: I have a small suggestion. You could try to use einops package for pytorch: https://einops.rocks, it makes your life easier when you are reshaping, permuting and transposing tensors and also your code is easier to read :)

Comment (JasiekKaczmarczyk) [reply models/VQVAE.py:238]: For example:
```python
flat_x = x.permute(0, 2, 3, 1).reshape(-1, self.embedding_dim)
```
changes to
```python
flat_x = einops.rearrange(x, "b c h w -> (b h w) c")
```

Comment (SamuelJanas) [reply models/VQVAE.py:238]: Will have to look more into it, after the `einops` commit I was getting size mismatches all around the board. Thank you for the suggestion but I'll be probably implementing that after some code clean up since it's a little mess for now 😅

**Comment by SamuelJanas** [comment]
Changed data loading to correctly normalize CIFAR dataset, VQ-VAE is working great. Here are the results after 10 epochs of training:
side_by_side
Images on the right (reconstructed) are only slightly blurry.

I am currently working on a fix regarding the other models as data loading might have been causing them to malfunction.
I will post the results of VAE performance on MNIST after I'm done fixing it.

**Comment by SamuelJanas** [comment]
Further fixes now allow to run all the configs I've prepared, there are still errors, namely with VAE. Visualization tool is either difffering colors or there is some problem with the model itself.

**Comment by SamuelJanas** [comment]
VAE is working fine with 1 channel data but breaks on CIFAR [1]. Googling for answer I found that it is possible to train it on CIFAR although the results are far from VQ-VAE, I would end the experiment with images for VAE unless we find it necessary?

[1]: The training seems to be working, although with not much success, it results in just gray background being displayed for everything. The issue lies somewhere in normalization I believe. With some tweaks I was able to make it display at least misty-shapes of the images. But honestly speaking I was repeating same debugging steps at this point and I feel a little stuck.

File: train.py [hunk 2]
```diff
@@ -32,10 +46,76 @@ def train(cfg: DictConfig, autoencoder: Autoencoder, loss_function: torch.nn.Mod
...
     else:
         train_autoencoder(cfg, autoencoder, loss_function, train_loader, test_loader)
 
 
+def train_vqvae(cfg: DictConfig, autoencoder: Autoencoder, loss_function: torch.nn.Module, train_loader, test_loader):
+    total_start_time = time.time()
+    process = psutil.Process()
+    best_train_loss = float("inf")
+    optimizer = torch.optim.Adam(autoencoder.parameters(), lr=cfg.train.lr)
+    for epoch in range(1, cfg.train.epochs + 1):
```
Comment (roszcz) [inline train.py:60]: Is there a reason not to start from 0? :)

File: utils/train_utils.py [hunk 3]
```diff
@@ -117,3 +123,69 @@ def test_epoch_ecg(autoencoder, test_loader, device, loss_function, verbose=True
         print(f"====> Test set loss: {avg_loss:.4f}")
 
     return avg_loss
+
+
+def train_epoch_vqvae(
+    autoencoder, train_loader, optimizer, device, log_interval, epoch, loss_function, beta, use_ema, best_train_loss
+):
```
Comment (roszcz) [inline utils/train_utils.py:130]:
```suggestion
def train_epoch_vqvae(
    autoencoder: VQVAE,
    train_loader: DataLoader,
    optimizer: torch.optimizer,
    device: str,
    log_interval: int,
    etc.
):
```

Type hints are always good, but here's one use case I wanted to highlight - I'm reading through your code and having the model class here allows me to jump quickly to its definition (most code editor can do that). 

If there's no type hint with the exact model class, I have to go backward in code through multiple function calls to figure out what exactly is the type of model allowed here.

File: models/VQVAE.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        self.decoder_layer = self.build_decoder(
+            embedding_dim,
+            num_hiddens,
+            num_downsampling_layers,
+            num_residual_layers,
+            num_residual_hiddens,
+        )
+
+    def build_encoder(self, in_channels, num_hiddens, num_downsampling_layers, num_residual_layers, num_residual_hiddens):
```
Comment (roszcz) [inline models/VQVAE.py:147]:
```suggestion
    def build_encoder(
        self,
        in_channels: int,
        num_hiddens: int,
        num_downsampling_layers: int,
        num_residual_layers: int,
        num_residual_hiddens: int,
    ) -> Encoder:
```

Making reading the code as straightforward as possible is our priority here, and type hint (including return) and per-line arguments is extremely helpful in this (for me at least).

You can always ask chat gpt to write type hints for you :sweat_smile:

**Review by roszcz — APPROVED** [review]
Looks good! Let's go ahead and merge this, I'll keep running experiments with those autoencoders, and if I find anything to improve we'll plan separate tasks :+1:

File: utils/data_loader.py [hunk 5]
```diff
@@ -2,9 +2,20 @@
 
 import torch
 from datasets import load_dataset
+from torchvision import transforms
```
Comment (roszcz) [inline utils/data_loader.py:5]: I think it's missing from `requirements.txt`

**Review by roszcz — APPROVED** [review]
Can you update the way logging works? Right now it prints a new line with every training log, and the output blows up vertically. Can you change it to some kind of tqdm-like progress bar? 

The crucial information to show on a training progress bar is:
- loss and optionally other metrics
- how far til the end (both epoch and training)
